pkg/providers/openai/chat.go

```diff
@@ -1,53 +1,262 @@
+	TopP             float64          `json:"top_p,omitempty" validate:"omitempty,gte=0,lte=1"`
+	MaxTokens        int              `json:"max_tokens,omitempty" validate:"omitempty,gte=0"`
+	N                int              `json:"n,omitempty" validate:"omitempty,gte=1"`
+	StopWords        []string         `json:"stop,omitempty"`
+	Stream           bool             `json:"stream,omitempty" validate:"omitempty, boolean"`
+	FrequencyPenalty int              `json:"frequency_penalty,omitempty"`
+	PresencePenalty  int              `json:"presence_penalty,omitempty"`
+	LogitBias        *map[int]float64 `json:"logit_bias,omitempty" validate:"omitempty"`
+	User             interface{}      `json:"user,omitempty"`
+	Seed             interface{}      `json:"seed,omitempty" validate:"omitempty,gte=0"`
+	Tools            []string         `json:"tools,omitempty"`
+	ToolChoice       interface{}      `json:"tool_choice,omitempty"`
+	ResponseFormat   interface{}      `json:"response_format,omitempty"`
+
+	// StreamingFunc is a function to be called for each chunk of a streaming response.
+	// Return an error to stop streaming early.
+	StreamingFunc func(ctx context.Context, chunk []byte) error `json:"-"`
+}
+
+// ChatMessage is a message in a chat request.
+type ChatMessage struct {
+	// The role of the author of this message. One of system, user, or assistant.
+	Role string `json:"role"`
+	// The content of the message.
+	Content string `json:"content"`
+	// The name of the author of this message. May contain a-z, A-Z, 0-9, and underscores,
+	// with a maximum length of 64 characters.
+	Name string `json:"name,omitempty"`
+}
+
+// ChatChoice is a choice in a chat response.
+type ChatChoice struct {
+	Index        int         `json:"index"`
+	Message      ChatMessage `json:"message"`
+	FinishReason string      `json:"finish_reason"`
+}
+
+// ChatUsage is the usage of a chat completion request.
+type ChatUsage struct {
+	PromptTokens     int `json:"prompt_tokens"`
+	CompletionTokens int `json:"completion_tokens"`
+	TotalTokens      int `json:"total_tokens"`
+}
+
+// ChatResponse is a response to a chat request.
+type ChatResponse struct {
+	ID      string        `json:"id,omitempty"`
+	Created float64       `json:"created,omitempty"`
+	Choices []*ChatChoice `json:"choices,omitempty"`
+	Model   string        `json:"model,omitempty"`
+	Object  string        `json:"object,omitempty"`
+	Usage   struct {
+		CompletionTokens float64 `json:"completion_tokens,omitempty"`
+		PromptTokens     float64 `json:"prompt_tokens,omitempty"`
+		TotalTokens      float64 `json:"total_tokens,omitempty"`
+	} `json:"usage,omitempty"`
+}
+
+// Chat sends a chat request to the specified OpenAI model.
+//
+// Parameters:
+// - payload: The user payload for the chat request.
+// Returns:
+// - *ChatResponse: a pointer to a ChatResponse
+// - error: An error if the request failed.
+func (c *ProviderClient) Chat() (*ChatResponse, error) {
```

Comment:
The UnifiedAPI data is assigned to the client during init then is accessed when chat is called.

Reply:
@mkrueger12 We would to avoid massive struct creation, so I would have one client that serves all requests with given configs. So we allocate memory for it once and it's working without further memory allocations. 

The client is simply inited with a config from the corresponding model pool it represents at service intialization and then we just call it methods with passing the unified schema for further processing. 

Besides saving some time not allocating memory, this would help us to keep state like connection pooling, possibly resiliency related things.

Does it make sense? 👀

Reply:
I kind of understand. So it would look something like this:

```go

newClient = Client(configFromPool) // what does this config look like? Is it from the config file?

resp, err = newClient.Chat(UnifiedAPIData) // this data matches discussion in GEP
```

Reply:
@mkrueger12 yes, that's where we should probably stop at for this PR. It's going to be wrapped into the pooling logic in the follow-up PRs.